Algorithm.Framework/Execution/VolumeWeightedAveragePriceExecutionModel.cs

```diff
@@ -0,0 +1,282 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Algorithm.Framework.Portfolio;
+using QuantConnect.Data;
+using QuantConnect.Data.Consolidators;
+using QuantConnect.Data.Market;
+using QuantConnect.Data.UniverseSelection;
+using QuantConnect.Indicators;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Algorithm.Framework.Execution
+{
+    /// <summary>
+    /// Execution model that submits orders while the current market price is more favorable that the current volume weighted average price.
+    /// </summary>
+    public class VolumeWeightedAveragePriceExecutionModel : IExecutionModel
+    {
+        private readonly PortfolioTargetCollection _targetsCollection = new PortfolioTargetCollection();
+        private readonly Dictionary<Symbol, SymbolData> _symbolData = new Dictionary<Symbol, SymbolData>();
+
+        /// <summary>
+        /// Gets or sets the maximum order quantity as a percentage of the current bar's volume.
+        /// This defaults to 0.01m = 1%. For example, if the current bar's volume is 100, then
+        /// the maximum order size would equal 1 share.
+        /// </summary>
+        public decimal MaximumOrderQuantityPercentVolume { get; set; } = 0.01m;
+
+        /// <summary>
+        /// Submit orders for the specified portolio targets.
+        /// This model is free to delay or spread out these orders as it sees fit
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance</param>
+        /// <param name="targets">The portfolio targets to be ordered</param>
+        public void Execute(QCAlgorithmFramework algorithm, IEnumerable<IPortfolioTarget> targets)
+        {
+            // update the complete set of portfolio targets with the new targets
+            _targetsCollection.AddRange(targets);
+
+            foreach (var target in _targetsCollection)
+            {
+                var symbol = target.Symbol;
+
+                // calculate remaining quantity to be ordered
+                var unorderedQuantity = OrderSizing.GetUnorderedQuantity(algorithm, target);
+
+                // fetch our symbol data containing our VWAP indicator
+                SymbolData data;
+                if (!_symbolData.TryGetValue(symbol, out data))
+                {
+                    continue;
+                }
+
+                // ensure we're receiving price data before submitting orders
+                if (data.Security.Price == 0m)
```

Comment:
Wonder if we can reuse the PreOrderChecks method. Wonder if these will grow over time

Reply:
There's another flag, `IsTradable` -- currently only a public get/set and we use it to set internal feeds to false, I could see it being a little smarter and also returning false when price is zero, and perhaps other things we may check (assuming they relate directly to the security, ofc)